File: aegis-node-agent/agent.js
const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

// The Radio Channel to the Central Brain
const BRAIN_URL = "http://localhost:8081/api/heal";
const TARGET_FILE = path.join(__dirname, 'target-server.js');

console.log("========================================");
console.log("🛡️ AEGIS NODE.JS SIDECAR: ONLINE");
console.log("========================================");

let currentServerProcess = null;

function startTargetServer() {
    console.log("[SIDECAR] Launching target-server.js in protective bubble...");

    currentServerProcess = spawn('node', [TARGET_FILE]);

    currentServerProcess.stdout.on('data', (data) => {
        process.stdout.write(`${data}`);
    });

    currentServerProcess.stderr.on('data', (data) => {
        const errorLog = data.toString();

        console.log("\n🚨 [CRASH DETECTED] Node.js Exception caught! 🚨");
        // Print just the first line of the error to keep the terminal clean
        console.log("-> " + errorLog.split('\n')[0]);

        // Kill the dead process completely before we start healing
        if (currentServerProcess) currentServerProcess.kill();

        initiateHealingSequence(errorLog);
    });
}

async function initiateHealingSequence(errorLog) {
    console.log("\n[SIDECAR] Reading broken source code...");
    const brokenCode = fs.readFileSync(TARGET_FILE, 'utf8');

    console.log("[SIDECAR] Transmitting JSON payload to Central Brain API (Port 8081)...");

    // We package the data exactly like the Python agent does!
    const payload = {
        language: "javascript",
        brokenCode: brokenCode,
        errorLog: errorLog
    };

    try {
        // We use Node's built-in fetch to talk to the Brain
        const response = await fetch(BRAIN_URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });

        const data = await response.json();

        if (data.status === 'success') {
            console.log("[SIDECAR] Patch received! Applying Hot-Swap...");

            // Overwrite the target server file with the AI's fix
            fs.writeFileSync(TARGET_FILE, data.fixedCode);

            console.log("========================================");
            console.log("⚡ RESURRECTION SEQUENCE INITIATED ⚡");
            console.log("========================================");

            startTargetServer(); // Reboot!
        } else {
            console.log("[CRITICAL] Brain returned an error: ", data.message);
        }
    } catch (error) {
        console.error("[CRITICAL] Failed to contact Central Brain. Is it running on Port 8081?");
    }
}

// Turn it on!
startTargetServer();

File: aegis-python-agent/agent.py
import subprocess
import requests
import json
import os

BRAIN_URL = "http://localhost:8081/api/heal"
TARGET_FILE = "target.py"

print("========================================")
print("🛡️ AEGIS PYTHON SIDECAR: ONLINE")
print("========================================")

def start_server():
    print("[SIDECAR] Launching target.py in protective bubble...")
    
    # Run the Python server and capture its output
    process = subprocess.Popen(
        ['python', TARGET_FILE], 
        stdout=subprocess.PIPE, 
        stderr=subprocess.PIPE,
        text=True
    )

    while True:
        # Read the normal logs
        output = process.stdout.readline()
        if output == '' and process.poll() is not None:
            break
        if output:
            print(output.strip())

    # If the process died, grab the error log!
    error_log = process.stderr.read()
    if error_log:
        print("\n🚨 [CRASH DETECTED] Python Traceback caught! 🚨")
        print("-> " + error_log.strip().split('\n')[-1]) # Print just the last line of the error
        initiate_healing_sequence(error_log)

def initiate_healing_sequence(error_log):
    print("\n[SIDECAR] Reading broken source code...")
    with open(TARGET_FILE, 'r') as file:
        broken_code = file.read()

    print("[SIDECAR] Transmitting JSON payload to Central Brain API (Port 8081)...")
    payload = {
        "language": "python",
        "brokenCode": broken_code,
        "errorLog": error_log
    }

    try:
        response = requests.post(BRAIN_URL, json=payload)
        data = response.json()

        if data['status'] == 'success':
            print("[SIDECAR] Patch received! Applying Hot-Swap...")
            
            # Overwrite the target.py file with the fixed code!
            with open(TARGET_FILE, 'w') as file:
                file.write(data['fixedCode'])
                
            print("========================================")
            print("⚡ RESURRECTION SEQUENCE INITIATED ⚡")
            print("========================================")
            start_server() # Reboot!
            
    except Exception as e:
        print(f"[CRITICAL] Failed to contact Central Brain. {e}")

# Turn it on!
start_server()